Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine+Demand.swift

```diff
@@ -92,10 +92,21 @@ extension HTTPRequestStateMachine {
                     return buffer
                 }
 
+            // For all the following cases, please note:
+            // Normally these code paths should never be hit. However there is one way to trigger
+            // this:
+            //
+            // If the connection to a server is closed, NIO will forward all outstanding
+            // `channelRead`s without waiting for a next `context.read` call. After all
+            // `channelRead`s are delivered, we will also see a `channelReadComplete` call. After
+            // this has happened, we know that we will get a channelInactive, which will fail the
+            // request. Since the request is doomed anyway, we can just drop the already buffered
+            // bytes here.
+
             case .waitingForRead,
                  .waitingForDemand,
                  .waitingForReadOrDemand:
-                preconditionFailure("How can we receive a body part, after a channelReadComplete, but no read has been forwarded yet. Invalid state: \(self.state)")
+                return nil
```

Comment:
Sorry, I know you asked about this, but I think we should preserve the bytes.

There are some rare circumstances where you might get multiple read cycles in this situation. NIO will spin reading from the socket until it's _empty_. If you have `maxMessagesPerRead` set to `1`, it's therefore possible to get to a situation where we might get a `channelReadComplete` and _then_ an `.end`. We should probably try not to pre-destroy the bytes here.

Reply:
Correct me if I'm wrong here, but we effectively don't destroy the bytes here, since we don't change our state. If we receive further `channelRead`s bytes will be added to the buffer (see method above). If we receive a final `.end` we will hand out all data.

Reply:
But at least the comment above says something different:
> Since the request is doomed anyway, we can just drop the already buffered bytes here.